### PR TITLE
Adds Tutorial

### PR DESCRIPTION
**Pull request by premalrupnur**
Yet to be completed

### CODE REVIEW

File: projects/machine_learning/Linear Regression.ipynb [hunk 1]
```diff
@@ -0,0 +1,154 @@
+{
```
Comment (oke-aditya) [inline projects/machine_learning/Linear Regression.ipynb:1]: <p>I guess pandas and numpy are not needed :) </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/madewithml/hacktoberfest/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (oke-aditya) [inline projects/machine_learning/Linear Regression.ipynb:1]: <p>Some clarification in cost function.</p><p> </p><p>We use Squared error loss function, not that cost function is called squared error.</p><p> </p><p>One of my proposal to explain ML models is as follows.</p><p> </p><p>1. Data</p><p>2 Task</p><p>3 Model</p><p>4 Loss function</p><p>5 Learning algorithm</p><p>6 Evaluation Metrics</p><p> </p><p>This leads to clean presentation of ML models.</p><p> </p><p>Have a look here :-) </p><p>https://medium.com/machine-learning-magazine/a-guide-to-supervised-learning-b9c227a71155</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/madewithml/hacktoberfest/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (premalrupnur) [reply projects/machine_learning/Linear Regression.ipynb:1]: Yes they aren't I would edit it

**Comment by premalrupnur** [comment]
> Hey, the draft looks fine, but we can structure it better.
> 
> Let me know your thoughts.
Yes once I complete it I would restructure it in better way

File: projects/machine_learning/Linear Regression.ipynb (hunk 1 above)
Comment (premalrupnur) [reply projects/machine_learning/Linear Regression.ipynb:1]: But as we are addressing a specific task that is regression how about explaining first then we would go for data part

**Comment by premalrupnur** [comment]
I hope this one has better structure than before

File: projects/machine_learning/Linear Regression.ipynb (hunk 1 above)
Comment (oke-aditya) [reply projects/machine_learning/Linear Regression.ipynb:1]: Brief overview about regression tasks is fine, just what we are going to do in this tutorial / blogpost.

File: projects/machine_learning/Linear Regression.ipynb [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
```
Comment (oke-aditya) [inline projects/machine_learning/Linear Regression.ipynb:1]: <p>Another quite popular metric is R2 score.</p><p> </p><p>This can be even obtained by <code>print(model.score)</code>  I think.</p><p> </p><p>Can you explain R2 score as well?</p><p> </p><p>MSE is fine, but in some cases we can use MAE (Mean absolute error as well)</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/madewithml/hacktoberfest/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (oke-aditya) [inline projects/machine_learning/Linear Regression.ipynb:1]: <p>I suggest to skip Bias Variance Trade-off here, this topic is actually bigger and has more cleaner statistical implication.</p><p> </p><p>We can cover in next notebook.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/madewithml/hacktoberfest/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (oke-aditya) [inline projects/machine_learning/Linear Regression.ipynb:1]: <p>We can keep Over fit and Underfit, but remove the terms <code>bias</code>  and <code>variance</code>  (These are statistical terms which need explainations)</p><p> </p><p>Can we inutiatvely explain Overfit and Underfit using 2 diagrams?</p><p> </p><p>And how regularization for Overfit will solve the problem?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/madewithml/hacktoberfest/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (oke-aditya) [inline projects/machine_learning/Linear Regression.ipynb:1]: <p>Skip this high-variance etc,</p><p> </p><p>This is certainly going to be one of the first tutorials a beginner will see. This definitely is hard to digest.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/madewithml/hacktoberfest/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (oke-aditya) [inline projects/machine_learning/Linear Regression.ipynb:1]: <p>Looks fine, but we need bit more content.</p><p> </p><p><code>Lasso</code>  is also quite popular.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/madewithml/hacktoberfest/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (premalrupnur) [reply projects/machine_learning/Linear Regression.ipynb:1]: > Can we intuitively explain Overfit and Underfit using 2 diagrams?

Ig I have used a image and explained the two terms

Comment (premalrupnur) [reply projects/machine_learning/Linear Regression.ipynb:1]: > And how regularization for Overfit will solve the problem?

May I use the simple example I had used for high variance ,to show how regularization would shrinks the slope , I mean creating 2 plots with random data  , in one we would plot 3 data points with no noise and no regularization  and in second plot the line using regularization which would then not perfectly pass through data

Comment (premalrupnur) [reply projects/machine_learning/Linear Regression.ipynb:1]: > Looks fine, but we need bit more content.

Ig explaining why regulariztion solves overfit would do the trick

Comment (premalrupnur) [reply projects/machine_learning/Linear Regression.ipynb:1]: Or else I can just say that the regularization terms contain slopes and are included in cost function so in order to reduce the cost function regularization term as to be reduced wich in turn would penalize the slopw

Comment (premalrupnur) [reply projects/machine_learning/Linear Regression.ipynb:1]: > MSE is fine, but in some cases we can use MAE (Mean absolute error as well)

would using MAE bring different(better) results to model than MSE?? , may I know how??
ig it would just cause difference to J v/s w plot but gradient descent should work fine(same) in both cases right???

Comment (oke-aditya) [reply projects/machine_learning/Linear Regression.ipynb:1]: No it depends, MAE is extremely useful in doing regression where we need to compare absolute values.

MAE is just metric it does not matter which algorithm you use while training.

Comment (premalrupnur) [reply projects/machine_learning/Linear Regression.ipynb:1]: > Can you explain R2 score as well?

is it ok if I skip this part I would write code for all three metrics mae,mseand r2 but I myself am not that clear with R2

Comment (premalrupnur) [reply projects/machine_learning/Linear Regression.ipynb:1]: ??

Comment (oke-aditya) [reply projects/machine_learning/Linear Regression.ipynb:1]: I leave this to you. The description should completely illustrate the idea of `overfit` vs `underfit` without being far too technical. Illustartions can be extremely helpful to illustrate.

File: projects/machine_learning/Linear Regression.ipynb [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
```
Comment (oke-aditya) [inline projects/machine_learning/Linear Regression.ipynb:1]: <p>Can we use <code>f</code>  strings</p><pre class="ql-syntax" spellcheck="false">print(f"mse train: {mse_train}")
</pre><p> </p><p>I think we can write mathematical formula of the MSE, MAE and R2.</p><p> </p><p>Also we can shift the below cell up.</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/madewithml/hacktoberfest/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (oke-aditya) [inline projects/machine_learning/Linear Regression.ipynb:1]: <p>Can we have code for these.</p><p> </p><p>There is a nice circle diagram that shows difference between these two (I think its in Hastie Book)</p><p> </p><p>Also in the last section you need to add references for images / resource you have used and would like the readers to refer to additionally.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/madewithml/hacktoberfest/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

**Review by oke-aditya — CHANGES_REQUESTED** [review]
Let me know your thoughts 😄

File: projects/machine_learning/Linear Regression.ipynb (hunk 3 above)
Comment (premalrupnur) [reply projects/machine_learning/Linear Regression.ipynb:1]: Yes I would cite the images , but I would like to ask that have you read entire Hastie book😂

Comment (premalrupnur) [reply projects/machine_learning/Linear Regression.ipynb:1]: Just want  a advice from you that if we have done courses should we read that book???

Comment (oke-aditya) [reply projects/machine_learning/Linear Regression.ipynb:1]: No I haven't read the book 😢 Though the course I learnt from had all content from it.
I Recommend to read only if you have time its pretty voluminous and very math heavy.

Comment (premalrupnur) [reply projects/machine_learning/Linear Regression.ipynb:1]: > Also we can shift the below cell up.

Why should this cell be shifted up because I have added it below the header "Assessing the goodness of fit", do you mean above this header???
